enhancements/catalog-source-config.md

```diff
@@ -0,0 +1,340 @@
+
+
+## Motivation
+
+This EP targets issue [#2452](https://github.com/operator-framework/operator-lifecycle-manager/issues/2452).
+Users would like to be able to override the pod spec of the CatalogSource pod
+created for `grpc` `CatalogSource`s. Especifically, they would like to be able to specify
+pod tolerations and node selectors.
+
+### Goals
+
+ * Update the `CatalogSource` spec to allow users to affect the pod spec of `grpc` `CatalogSource` pods 
+
+### Non-Goals
+
+ TBD
+
+## Proposal
+
+### Current State
+
+The [current](https://github.com/operator-framework/api/blob/a80624eab36b2005fa5d84c9b539bb4817e44daa/crds/operators.coreos.com_catalogsources.yaml#L54) `CatalogSource.spec` accepts three difference values for the `spec.sourceType`: `grpc`, `configmap`, and `internal`. Depending on the value, different optional attributes are used to deploy the underlying infrastructure to serve the catalog content. For instance, if `spec.sourceType` is set to `grpc`, the user can set either the `spec.address` or `spec.image`. If both are set, `spec.image` takes precedence. If `spec.sourceType` is set to `configmap`,
+`spec.configMap` gets used, but `spec.address` and `spec.image` get ignored. Since there could be attributes defined that just get ignored by the controller, it could be hard for a user to know what is happening given just a `CatalogSource` CR and no documentation.
+
+### Options
+
+Three different options are outlined below. The first two would require breaking changes to the API, while the third would be the "minimal" optional. The path taken likely depends on the whether we are planning for a new `CatalogSource` CRD version in the near future or not.
+
+#### Union Types
+
+We use a union type to define the type and configuration for the `CatalogSource`. This can be realised with the [`oneOf`](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/) keyword supported in OpenAPI v3 CRD schemas. The schema would then look something like:
+
+```yaml
+spec:
+  type: object
+  required:
+    - sourceConfig
+  properties:
+    # Deprecate 
+    # image: ... 
+    # address: ... 
+    # configMap: ...
+    # sourceType: grpc | configMap | internal
+    
+    # Keep as is:
+    description: ...
+    displayName: ...
+    icon: ...
+    priority: ...
+    publisher: ...
+    secrets: ...
+
+    # Add:
+    # Note: 'internal' type is being deprecated so it gets no config structure
+    grpcPodSource:
+      podSpec:  # a subset (or the whole of) corev1.PodSpec
+        image: ... # required
+        # all other attributes are optional
+        # and override the default catalog source pod spec
+        ...
+      ...
+    grpcServiceSource:
+      address: ...
+      ...
+    configMapSource:
+      configMap: ...
+      ... 
+  oneOf:
+    - required: ["grpcPodSource"]
+    - required: ["grpcServiceSource"]
+    - required: ["configMapSource"]
+                  
+``` 
```

Comment:
I am missing the details of what would be added to the CRD. Looking at Prometheus it can be quite extensive:
https://github.com/prometheus-operator/prometheus-operator/blob/main/jsonnet/prometheus-operator/prometheuses-crd.json#L126
Cluster-Logging has a far more limited configurability

Reply:
It really depends how deep we want to go. As I see it, we have a couple of options:
1. use `corev1.PodSpec` 
2. start small and just have a map for the node-selector and a []corev1.Toleration for the tolerations

E.g. something like:

```go
// NodeSelector is a selector which must be true for the pod to fit on a node.
// Selector which must match a node's labels for the pod to be scheduled on that node.
// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
// +optional
NodeSelector map[string]string `json:"nodeSelector,omitempty"`

// Tolerations are the catalog source's pod's tolerations.
// +optional
Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
```

This will be the next decision to make. At the moment I more concerned with the approach. But fair call out!

Reply:
But yeah, what ever we do, it will add a lot of bits to the CRD. Probably something similar to the subscriptionConfig https://github.com/operator-framework/api/blob/master/crds/operators.coreos.com_subscriptions.yaml#L62